### ISSUE TITLE
FetchResult type unsoundness

### ISSUE THREAD

**Issue by infogulch** [apollographql/apollo-link#616]
`FetchResult<T>` appears to have an error in the way it's generic parameters are defined and used, affecting its' ability to be typechecked correctly.

Looking at `apollo-client/core/QueryManager.ts` the definitions of [`query<T>`](https://github.com/apollographql/apollo-client/blob/9ebbb61fb1061f56edf0bedb965c9954f75afead/packages/apollo-client/src/core/QueryManager.ts#L657) and [`mutate<T>`](https://github.com/apollographql/apollo-client/blob/9ebbb61fb1061f56edf0bedb965c9954f75afead/packages/apollo-client/src/core/QueryManager.ts#L119) appear to use their type parameter in a similar way: to declare the type of the `data` property that is passed to its' subscription handler. This works out in the case `ApolloQueryResult<T>`, this is untrue for `FetchResult<T>` which I believe is a bug.

ApolloQueryResult is defined in [apollo-client/src/core/types.ts](https://github.com/apollographql/apollo-client/blob/9ebbb61fb1061f56edf0bedb965c9954f75afead/packages/apollo-client/src/core/types.ts#L16) as:
``` typescript
export declare type ApolloQueryResult<T> = {
    data: T;
    // snip
};
```
Note that the first type parameter declares the type of the `data` property.

FetchResult is defined in [apollo-link/src/types.ts](https://github.com/apollographql/apollo-link/blob/623b04cd145b68e2d455f3db6e971dac8bc7d738/packages/apollo-link/src/types.ts#L23) as:
``` typescript
export type FetchResult<
  C = Record<string, any>,
  E = Record<string, any>
> = ExecutionResult & {
  extensions?: E;
  context?: C;
};
```

Note that the first type parameter C is used to declare the type of `context`, and the type of ExecutionResult's `data` property is left undeclared. Comparing to ApolloQueryResult, one would expect the first type parameter to declare the type of the data property.

I believe `FetchResult` should instead be defined as:
``` typescript
export type FetchResult<
  T = any,
  C = Record<string, any>,
  E = Record<string, any>
> = ExecutionResult & {
  data?: T;
  extensions?: E;
  context?: C;
};
```

**Comment by kamilkisiela** [comment]
@jbaxleyiii @evans I would create a PR for it, to use `T,C,E`, where `T` is for `data` but I need to know If it's like this by design. If not then probably introducing a new Interface that extends `FetchResult` and has three generic types, would be a good idea.

Would be super nice to solve it somehow :)

**Comment by infogulch** [comment]
>  I need to know If it's like this by design

Yeah that was my hangup and why this isn't a pull request already. To me it really *looks* like this is a mistake and the first generic parameter *should be* type of the data property, but since the type system starts petering out in the internals I'm not 100% sure.

On a side note I've been going through adding generic types throughout the apollo repos api (-link, -client, -angular, with some integration in -codegen) and that has reaffirmed my thoughts about FetchResult, and should really help with type-safety throughout the api. That will probably be a pull request in the next week or so.

**Comment by ColCh** [comment]
Guys, how to hotfix this until it would be resolved? Only by using forks?

**Comment by yezarela** [comment]
any update on this?

**Comment by jhliberty** [comment]
Also very interested in an update on this

**Comment by ColCh** [comment]
I have resolved this by using mutation component - it types greatly. Link to docs https://www.apollographql.com/docs/react/essentials/mutations.html#basic

However, `any` in data was not main reason for using component

**Comment by kgtkr** [comment]
Why data is optional(`?`)?
`data?: { [key: string]: any }`

**Comment by alexk111** [comment]
Any progress on this? What's blocking the update?

**Comment by michaelbromley** [comment]
A more recent version of `@types/graphql` [makes the `ExecutionResult` type generic](https://github.com/DefinitelyTyped/DefinitelyTyped/commit/deb7d8beaff5d43da749c5bc0f2020cdcaaabb6b#diff-1a92e4f388a3da0eb35da5e193f02172)

So another solution would be to update the graphql typings and then define `FetchResult` as:

```ts
export type FetchResult<
  C = Record<string, any>,
  E = Record<string, any>
> = ExecutionResult<C> & {
  extensions?: E;
  context?: C;
};
```
However, I also noticed that in the mean time for some reason `@types/graphql` was made a devDependency (#735). I'm not sure of the reasoning behind that, but it you were to opt for the solution I propose here, then `"@types/graphql": "0.13.4",` would have to be at least a peerDependency, if not a dependency.

**Comment by kamilkisiela** [comment]
@michaelbromley  I think it's worth to think about it.

Few things:
- having `ExecutionResult<T>` would introduce a breaking change
- by making it to be a dependency might make a noise when you have a project that uses different version of `@types/graphql` and types won't match
- making it a peerDependency would notify user that we require something different here

My solution would be to just change the `FetchResult`, I don't think anyone uses C and E anyway. If so, then I'm guessing it's just 10 people. 
By introducing a third generic type in FetchResult would also bring a breaking change. UI libraries would have to bump requirements of apollo-client, apollo-link etc.

**Comment by michaelbromley** [comment]
@kamilkisiela good points. In my case, I came across this issue when I tried to use the `MutationUpdaterFn` in `apollo-client` and I noticed that the `data` property in the result was no longer type-safe.

So for that situation, simply adding a `data?: C` field to `FetchResult` would be sufficient. From my research I also think the C and E type args are not really used too.

**Comment by victorandree** [comment]
@michaelbromley @kamilkisiela 

**TL;DR:**

> `@types/graphql` should probably be a peer dependency, but should not be `dependency`.

---

I've been involved in some discussions about `@types/graphql` in this package. For reference, #634 gives background on the `@types/graphql` dependency. My reasoning there was: 

> This package should not expose @types/graphql as a dependency at all, since it doesn't re-export anything from graphql (if it did, graphql should be a production dependency and then @types/graphql should be a production dependency).

I would modify that statement to say that `graphql` and `@types/graphql` should be peer dependencies if

1. This package requires but does not bundle/vendor graphql (it does require graphql but does not bundle it, graphql is a peer dependency), and
2. This package makes parts of `graphql` or `@types/graphql` part of its interface by e.g. accepting or returning objects from those packages. I don't know if "it doesn't re-export anything from graphql" is (or was) true, but if it isn't, then `@types/graphql` should be a peer dependency. `ExecutionResult` is from `@types/graphql` so it seems like types from `@types/graphql` are part of this library.

[Basically, if a library exposes another library but does not vendor it, it should be a peer dependency](https://stackoverflow.com/questions/26737819/why-use-peer-dependencies-in-npm-for-plugins). The downside to making `@types/graphql` a `peerDependency` is that non-TypeScript users will be nagged about it. That's a decent argument for keeping it to `devDependencies`.

**Comment by kevinthecity** [comment]
> So for that situation, simply adding a data?: C field to FetchResult would be sufficient. From my research I also think the C and E type args are not really used too.

^ Anyone know a way to patch this without waiting for a lib update?

**Comment by rgolea** [comment]
Hey guys! Any news on this? I saw the PR but I am not sure this works.
